### ISSUE TITLE
Adjust neural network hyper-parameters

### ISSUE THREAD

**Issue by Chicoryn** [Chicoryn/dream-go#5]
The regularization coefficient _1e-4_ that DeepMind recommended was for their 40 layers architecture. We use the 20 layers architecture and therefore should have a larger regularization coefficient.

While we are at it, change the learning rate schedule and optimizer to Adam so that we can reach a local minimum faster as it takes ages on my non-supercomputer to train today.

- [x]  Add support for training on multiple GPU's.
- [x]  Change the learning rate schedule to something more reasonable.
- [x]  Change the optimizer to Adam.
- [x]  Adjust regularization constant.

**Comment by Chicoryn** [comment]
The 40 layer tower that DeepMind used had 11,895,775 weights and a regularization coefficient of 1e-4, while our 20 layer tower has 5,997,535 so our regularization coefficient should be approximately 2e-4.

```
0.0001 * (11,895,775 / 5,997,535) = 0.000198344
```

## 40 layers weights
```
3 * 3 * 34 * 128
+ 39 * (2 * 3 * 3 * 128 * 128)
+ 128 * 1 + 361 * 256 + 256 + 256 + 1
+ 128 * 2 + 722 * 362 + 362
= 11895775
```

## 20 layers weights
```
3 * 3 * 34 * 128
+ 19 * (2 * 3 * 3 * 128 * 128)
+ 128 * 1 + 361 * 256 + 256 + 256 + 1
+ 128 * 2 + 722 * 362 + 362
= 5997535
```

**Comment by Chicoryn** [comment]
In addition to treating DeepMind as an oracle, we might also wish to play with the regularization coefficient a bit ourselves since we can set it higher to avoid overfitting. This is especially useful because we are severely lacking in data compared to other similar projects.

I suggest training a 20 layer tower with the following coefficients on human games and observing their tournament performance.

* **1e-4** This is what the current network is trained with.
* **2e-4** As suggested by the previous post.
* **1e-3** To represent an extreme coefficient.

**Comment by Chicoryn** [comment]
I suggest the learning schedule in Table 1 based on empirical data gathered from training the 8 layer tower and previous 20 layer towers.

This schedule takes about 22 hours to fully train on my computer with two GTX 1080 Ti's.

<table>
<caption align="bottom">Table 1: Learning rate</caption>
<thead><tr><th>Step</th><th>Learning Rate</th></tr></thead>
<tbody>
<tr><td>0</td><td>3e-3</td></tr>
<tr><td>12,000</td><td>1e-3</td></tr>
<tr><td>27,000</td><td>3e-4</td></tr>
<tr><td>45,000</td><td>1e-4</td></tr>
<tr><td>66,000</td><td>3e-5</td></tr>
<tr><td>90,000</td><td>1e-5</td></tr>
</tbody>
</table>

**Comment by Chicoryn** [comment]
The tournament and testing accuracy suggest seems to correlate to an hierarchy where `1e-4` > `2e-4` > `1e-3`. It is unclear if, given an equal number of training steps the same result would occur.

In the meantime I suggest we keep the currently trained network as the _state of the art_ but keep the regularization coefficient for any future networks trained on self-play. The reasoning for this is that we will have issues with weights being overfit and the loss in playing strength from this larger regularization coefficient is not too large (even if it is non-trivial).

Would however recommend that when training on human data, we decrease the regularization coefficient to `1e-4`.
